src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp

```diff
@@ -521,19 +358,19 @@ AlgorithmIdentifier Dilithium_PublicKey::algorithm_identifier() const {
 }
 
 OID Dilithium_PublicKey::object_identifier() const {
-   return m_public->mode().oid();
+   return m_public->mode().mode().object_identifier();
 }
 
 size_t Dilithium_PublicKey::key_length() const {
    return m_public->mode().public_key_bytes();
 }
```

Comment:
I get confused by this functions every single time...
```cpp
/**
 * Return an integer value best approximating the length of the
 * primary security parameter. For example for RSA this will be
 * the size of the modulus, for ECDSA the size of the ECC group,
 * and for McEliece the size of the code will be returned.
 */
virtual size_t key_length() const = 0;
```
I'm not sure which value fits best here. We should, at least, provide some size in bits.
Same for Kyber.

Reply:
For Kyber, we could return the ML-KEM-x values: i.e. {512, 768, 1024}. What do these even mean though? My best guess: `k*N` as the coefficients in the private vector `s`?! At least that has some foundation in the names of the specified parameter sets, but its not a bit length. We could multiply those by 12 as the bit-length of Q, but that will result in fairly arbitrary values {6144, 9216, 12288}.

For Dilithium, we could go for the bit lengths of `s1` and `s2` perhaps?

@randombit Do you have any preference?

Reply:
When looking at where `key_length` is used at all, it seems that it should be some metric beside `estimated_strength` to identify specific instances of an algorithm while being generic enough to be a number. Also, the higher the number, the higher the security. Therefore, I would guess everything with these properties should be sufficient, even {1,2,3}. Therefore, I think {512, 768, 1024} should suffice. But I'll submit to the judgment of our Botan master ;)

Reply:
@FAlbertDev has the right idea about the intent. It doesn't really have anything to do with the length of the key in bytes in any direct way (after all a ECDSA P-256 public key is 64 bytes being two field elements, but saying the key length of ECDSA P-256 is 512 bits is not going to illuminate the situation!).

For Kyber I see quite commonly ML-KEM-{512,768,1024} so that seems the obvious choice. NIST uses the same naming.

For Dilithium I have seen less press. The NIST spec (as well as https://www.ietf.org/id/draft-ietf-lamps-dilithium-certificates-03.html) uses 44, 65, and 87.  TBH I have no idea where those numbers come from, but it's as good as anything.

Reply:
4x4, 6x5 and 8x7 are the dimensions of the generated public matrix A for the Dilithium instances.